wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/DevFileService.java

```diff
@@ -0,0 +1,148 @@
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.dataformat.yaml.YAMLFactory;
+import java.io.IOException;
+import javax.inject.Inject;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.Response;
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.core.ConflictException;
+import org.eclipse.che.api.core.NotFoundException;
+import org.eclipse.che.api.core.ServerException;
+import org.eclipse.che.api.core.ValidationException;
+import org.eclipse.che.api.core.rest.Service;
+import org.eclipse.che.api.devfile.model.Devfile;
+import org.eclipse.che.api.workspace.server.WorkspaceLinksGenerator;
+import org.eclipse.che.api.workspace.server.WorkspaceManager;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceConfigImpl;
+import org.eclipse.che.api.workspace.server.model.impl.WorkspaceImpl;
+import org.eclipse.che.api.workspace.shared.dto.WorkspaceDto;
+import org.eclipse.che.commons.env.EnvironmentContext;
+
+@Path("/devfile")
+public class DevFileService extends Service {
+
+  private WorkspaceLinksGenerator linksGenerator;
+  private DevFileSchemaValidator schemaValidator;
+  private WorkspaceManager workspaceManager;
+  private ObjectMapper objectMapper;
+  private DevFileConverter devFileConverter;
+
+  @Inject
+  public DevFileService(
+      WorkspaceLinksGenerator linksGenerator,
+      DevFileSchemaValidator schemaValidator,
+      WorkspaceManager workspaceManager) {
+    this.linksGenerator = linksGenerator;
+    this.schemaValidator = schemaValidator;
+    this.workspaceManager = workspaceManager;
+    this.objectMapper = new ObjectMapper(new YAMLFactory());
+    this.devFileConverter = new DevFileConverter();
+  }
+
+  //  Creates a workspace by providing the url to the repository
+  //  Initially this method will return
+  //  empty workspace configuration. And will start che-devfile-broker on a background to clone
+  //  sources and get devfile.
+  //  @POST
+  //  @Produces(APPLICATION_JSON)
+  //  public Response create(@QueryParam("repo_url") String repo_url){
+  //  }
+  //
+  //
+
+  /**
+   * Creates workspace from provided devfile
+   *
+   * @param data devfile content
+   * @param verbose return more explained validation error messages if any
+   * @return created workspace configuration
+   */
+  @POST
+  @Consumes({"text/yaml", "text/x-yaml", "application/yaml"})
+  @Produces(APPLICATION_JSON)
+  public WorkspaceDto createFromYaml(String data, @QueryParam("verbose") boolean verbose)
+      throws ServerException, ConflictException, NotFoundException, ValidationException,
+          BadRequestException {
+
+    Devfile devFile;
+    WorkspaceConfigImpl workspaceConfig;
+    try {
+      schemaValidator.validateBySchema(data, verbose);
```

Comment:
1. You parse input data twice, the first time during validation and the second time during Parsing object to Devfile. Maybe there is a possible way to avoid doing it.
2. IMHO Service would look better if we move serialization/deserizalization logic to `MessageBodyWriter`, `MessageBodyReader` like we do with our DTO's and CheJsonProvider. You can take a look how it is implemented here https://github.com/eclipse/che/blob/07263f1e30089689d71b057f747a44a29283e3c4/wsagent/che-core-api-debug/src/main/java/org/eclipse/che/api/debugger/server/DebuggerActionProvider.java#L52

Reply:
1 st - fixed;
2 nd - dont think it's needed right now, since we constructing always the same object, and its creation is pretty simple. (in your example, different objects created depending of souce data).
We may probably need that when different versions of devfile will be introduced. (to know which object to create) so propose to review it later.